### PR TITLE
Fix primary assembly heuristic + clean up duplicate code

### PR DESCRIPTION
**Pull request by vzarytovskii**
## Description

This fixes primary assembly check by:
1. Using Equals, which will use span helpers on latest runtime
2. Fixing the logic (before it used `Compare <> 0 || ...` now its just `Equals ||`, which will exit on the first match, as opposed to checking everything.
3. It will also now inline correctly (see stacks in the attached profiles.

It also gets rid of specific cache for int32 in the ilreader, as well as hides stats under define.

FSI startup is a bit faster on M2, should be more noticeable on older/slower machines:

Before:
image

After:
image

Before: [dotnet_trace_before.speedscope.json](https://github.com/dotnet/fsharp/files/15326074/dotnet_trace_before.speedscope.json)
After: [dotnet_trace_after.speedscope.json](https://github.com/dotnet/fsharp/files/15326071/dotnet_trace_after.speedscope.json)

### CODE REVIEW

**Comment by vzarytovskii** [comment]
Also a note - we spend quite some time generating PDB in all scenarios, even though it will never be used.

**Review by KevinRansom — APPROVED** [review]

**Review by KevinRansom — APPROVED** [review]

File: src/Compiler/AbstractIL/ilread.fs [hunk 1]
```diff
@@ -926,58 +926,33 @@ type GenericParamsIdx = GenericParamsIdx of numTypars: int * TypeOrMethodDefTag
 // Polymorphic caches for row and heap readers
 //---------------------------------------------------------------------
 
-let mkCacheInt32 lowMem _inbase _nm _sz =
+let mkCacheGeneric lowMem _inbase _nm (sz: int) =
     if lowMem then
-        (fun f x -> f x)
+        id
```
Comment (vzarytovskii) [inline src/Compiler/AbstractIL/ilread.fs:931]: wait, I didn't intend this change

**Review by T-Gro — APPROVED** [review]

**Comment by vzarytovskii** [comment]
> > Also a note - we spend quite some time generating PDB in all scenarios, even though it will never be used. Around 1/3rd of total assembly generation time, for example on my M2 it takes 50ms to generate metadata and additional 24ms to generate pdb.
> 
> Should we create an issue about it to keep track?

I did already. Not sure why it didn't link it here.

**Pull request merged**